File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
import itertools

def solution(orders, course):
    answer = []
    
    for c in course:
        order_comb = {}
        for order in orders:
            for comb in itertools.combinations(sorted(order), c):
                sorted_comb = ''.join(comb)
                if sorted_comb in order_comb:
                    order_comb[sorted_comb] += 1
                else:
                    order_comb[sorted_comb] = 1
        
        if order_comb:
            max_count = max(order_comb.values())
            if max_count >= 2:
                answer += [k for k, v in order_comb.items() if v == max_count]
                    
    return sorted(answer)

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {
    Map<String, Integer> map = new HashMap<>();

    public String[] solution(String[] orders, int[] course) {
        List<String> answer = new ArrayList<>();

        for (int len : course) {
            map.clear();
            for (String order : orders) {
                char[] arr = order.toCharArray();
                Arrays.sort(arr);
                combination(arr, new StringBuilder(), 0, len);
            }

            int max = 0;
            for (int v : map.values()) {
                max = Math.max(max, v);
            }

            if (max >= 2) {
                for (String key : map.keySet()) {
                    if (map.get(key) == max) {
                        answer.add(key);
                    }
                }
            }
        }

        Collections.sort(answer);
        return answer.toArray(new String[0]);
    }

    private void combination(char[] arr, StringBuilder sb, int idx, int targetLen) {
        if (sb.length() == targetLen) {
            map.put(sb.toString(), map.getOrDefault(sb.toString(), 0) + 1);
            return;
        }
        for (int i = idx; i < arr.length; i++) {
            sb.append(arr[i]);
            combination(arr, sb, i + 1, targetLen);
            sb.deleteCharAt(sb.length() - 1);
        }
    }
}
